zerver/lib/cache.py

```diff
@@ -581,11 +574,12 @@ def delete_user_profile_caches(user_profiles: Iterable["UserProfile"]) -> None:
 
     keys = []
     for user_profile in user_profiles:
-        keys.append(user_profile_by_email_cache_key(user_profile.delivery_email))
         keys.append(user_profile_by_id_cache_key(user_profile.id))
         for api_key in get_all_api_keys(user_profile):
             keys.append(user_profile_by_api_key_cache_key(api_key))
         keys.append(user_profile_cache_key(user_profile.email, user_profile.realm))
+        if user_profile.email != user_profile.delivery_email:
+            keys.append(user_profile_cache_key(user_profile.delivery_email, user_profile.realm))
```

Comment:
I don't think this is correct -- why would flushing with `user_profile_cache_key`  depend on whether `user_profile.email` is the same as `user_profile.delivery_email`?  My expectation is that a given cache should only be by a single version of email -- `.delivery_email` or not, with no mixed data.  My belief is `user_profile_cache_key` is meant to be for `email`, and `delivery_email` is not meant to be a cache key?  That's at least the situation with `get_user_by_delivery_email` not having a cache.

Reply:
It should be clearer now with  `actions: Change caching of create_mirror_user_if_needed.` extracted. The commit message has more elaboration - `delivery_emails` will also happen in the cache. Or is that too much of a complication and we should use  a new cache key for the create_mirror_user function's purposes?

Reply:
We never want to pollute a cache with different types of keys -- it can produce surprising unexpected behavior.  As an example, we at some point had a really tricky to debug issue (I think just in development for a PR) caused by reusing the same cache key for both active users and inactive users -- the bug being that sometimes you'd do a fetch for just active users and have an inactive user return, because they were in the cache!

It's much easier to reason about things and avoid potentially really subtle bugs if we just never creating reuse of cache keys.